src/clims/handlers.py

```diff
@@ -21,10 +31,154 @@ class HandlerContext(object):
     organization is etc.
     """
 
-    def __init__(self, organization):
+    def __init__(self, organization=None):
         self.organization = organization
 
 
+class HandlerManager(object):
+    """
+    Manages plugin defined handlers.
+    """
+
+    def __init__(self):
+        # Initialize the handlers dictionary with the types of baseclasses we can find:
+        self.handlers = self.find_handler_baseclasses()
+        logger.debug("Handler dictionary initialized: '{}'".format(self.handlers.keys()))
+
+    def add_handler_implementation(self, handler_type, implementation):
+        self.handlers[handler_type].add(implementation)
+
+    def remove_implementations(self):
+        for key in self.handlers:
+            self.handlers[key] = set()
+
+    def find_handler_baseclasses(self):
+        """
+        Finds all baseclasses of `clims.handlers.Handler`.
+        """
+        handlers = dict()
+        handler_subclasses = Handler.__subclasses__()
+        for subclass in handler_subclasses:
+            logger.info("Found handler type {}".format(subclass))
+            handlers[subclass] = set()
+        return handlers
+
+    def handler_count(self, cls):
+        return len(self.handlers[cls])
+
+    def require_single_handler(self, cls):
+        handlers = self.handlers[cls]
+        count = len(handlers)
+        if count == 0:
+            raise RequiredHandlerNotFound("No handler that implements '{}' found".format(cls))
+        elif count > 1:
+            # TODO: New type or change the type name
+            raise RequiredHandlerNotFound("Too many registered handlers found for '{}'".format(cls))
+
+    def require_handler(self, cls):
+        handlers = self.handlers[cls]
+        count = len(handlers)
+        if count == 0:
+            raise RequiredHandlerNotFound("No handler that implements '{}' found".format(cls))
+
+    def handle(self, cls, context, required, *args, **kwargs):
+        """
+        Runs all handlers registered for cls in sequence. *args are sent to the handler as arguments.
+
+        Returns a list of the handlers that were executed
+        """
+        logger.info("Handling action {}".format(cls))
+
+        ret = list()
+        if required:
+            self.require_handler(cls)
+        handlers = self.handlers[cls]
+        logger.debug("Handlers found for '{}': '{}'".format(cls, handlers))
+
+        for handler in handlers:
+            # TODO: the plugins manager should have an instance of the app
```

Comment:
I didn't do it there because I wanted to make sure that the PluginManager was made in the ApplicationService first, to make the lifetime clear. That is, the PluginManger is being used as a global instance which might be created before the ApplicationService.

In the PR I'm working on now, I fix this, by ensuring that the PluginManager is just one of the services created by the ApplicationService and it's a lot cleaner and easier to test.

Reply:
... so yeah, I broke it out so this could go in sooner.